docs/torch/distrib.rst

```diff
@@ -0,0 +1,353 @@
+    ...         # to track labels from the seed array. We do it with argsort to get unique elements
+    ...         # in the original order and broadcasting to get inverse indices
+    ...         unique_nodes, first_occurrence_indices = np.unique(edges, return_index=True)
+    ...         sort_order = np.argsort(first_occurrence_indices)
+    ...         ordered_unique_nodes = unique_nodes[sort_order]
+    ...         broadcasted_comparison = edges[:, None] == ordered_unique_nodes
+    ...         inverse_indices = np.argmax(broadcasted_comparison, axis=1)
+    ...
+    ...         edge_len = len(edges) // 2
+    ...         col = inverse_indices[:edge_len]
+    ...         row = inverse_indices[edge_len:]
+    ...         return ordered_unique_nodes, col, row
+    ...
+    ...     def query(self, _: int) -> tuple:
+    ...         edges = self.graph.sample_edges(
+    ...             self.batch_size,
+    ...             np.array([self.seed_edge_type], dtype=np.int32),
+    ...             strategy=SamplingStrategy.Weighted,
+    ...         )
+    ...         src = edges[:, 0]
+    ...         dst = edges[:, 1]
+    ...         num_pos = src.shape[0]
+    ...         num_neg = num_pos
+    ...         neg_edges = self.graph.sample_nodes(
+    ...             2 * num_neg,
+    ...             np.array(self.neg_node_type, dtype=np.int32),
+    ...             SamplingStrategy.Weighted,
+    ...         )[0]
+    ...         seed = np.concatenate(
+    ...             [src, neg_edges[:num_neg], dst, neg_edges[num_neg:]], axis=0
+    ...         )
+    ...         edge_label = np.zeros(num_pos + num_neg)
+    ...         edge_label[:num_pos] = 1
+    ...         seed, inverse_seed = np.unique(seed, return_inverse=True)
+    ...         edge_label_index = inverse_seed.reshape((2, -1))
+    ...         nodes, cols, rows = self._make_edge_index(seed)
+    ...         feats = self.graph.node_features(
+    ...             nodes,
+    ...             np.array([[1, self.feature_dim], [0, self.label_dim]], dtype=np.int32),
+    ...             np.float32,
+    ...         )
+    ...
+    ...         return (
+    ...             feats[:, : self.feature_dim],
+    ...             cols,
+    ...             rows,
+    ...             edge_label_index,
+    ...             edge_label,
+    ...             feats[:, self.feature_dim :],
+    ...             nodes,
+    ...         )
+
+Experiment configuration
+========================
+Next step is to configure our experiment
+.. code-block:: python
+
+    >>> from deepgnn.graph_engine.snark.distributed import Server, Client as DistributedClient
+    >>> config = {
+    ...     "data_dir": "/tmp/ppi",  # graph.data_dir(),
+    ...     "device": torch.device("cpu"),
+    ...     "num_epochs": 2,
+    ...     "num_nodes": 56944,  # graph.NUM_NODES,
+    ...     "address": "localhost:9999",
+    ...     "feature_idx": 1,
```

Comment:
Also I thought we decided ray config wasn't great for examples and we should just hardcode the values instead?

Reply:
There will be way more hardcoded values involved here duplicating in multiple datasets we create on lines 271. I went back and forth about it. Current code looked a bit better. Neither is good though, so I am open to better solutions.